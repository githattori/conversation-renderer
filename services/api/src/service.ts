import { applyGuardRails } from './guards.js';
import { buildSystemPrompt } from './presets.js';
import { buildContext, buildContract, computeDiffPatch } from './pipeline/dslBuilders.js';
import { extractEntities } from './pipeline/entityExtractor.js';
import { inferDiagramType } from './pipeline/diagramType.js';
import { LLMEntityExtractor } from './llm/entityExtractor.js';
import { SessionStore } from './sessionStore.js';
import {
  DiagramContract,
  DiagramContext,
  DiagramFormat,
  DiffPatch,
  MessageInput,
  RolePreset,
  Session,
} from './types.js';

interface ProcessResult {
  contract: DiagramContract;
  diff: DiffPatch;
  systemPrompt: string;
}

export class DiagramService {
  private store: SessionStore;
  private contexts = new Map<string, DiagramContext>();
  private llmExtractor: LLMEntityExtractor;

  constructor(store?: SessionStore, llmExtractor?: LLMEntityExtractor) {
    this.store = store ?? new SessionStore();
    this.llmExtractor = llmExtractor ?? new LLMEntityExtractor();
  }

  createSession(role: RolePreset): Session {
    return this.store.create(role);
  }

  getSession(sessionId: string): Session | undefined {
    return this.store.get(sessionId);
  }

  private setContext(sessionId: string, context: DiagramContext): void {
    this.contexts.set(sessionId, context);
  }

  private getContext(sessionId: string): DiagramContext | undefined {
    return this.contexts.get(sessionId);
  }

  async processMessage(input: MessageInput): Promise<ProcessResult> {
    const session = this.getSession(input.sessionId);
    if (!session) {
      throw new Error('Session not found');
    }
    this.store.appendMessage(input);

    const history = this.store.getHistory(input.sessionId);
    const combinedText = history.map((item) => item.content).join('\n');

    const llmExtraction = await this.llmExtractor.extract(
      history.map((item) => ({ role: 'user', content: item.content })),
      session.role,
    );

    const entities = llmExtraction?.entities?.length ? llmExtraction.entities : extractEntities(combinedText);
    const diagramType = llmExtraction?.diagramType ?? inferDiagramType(entities, combinedText);
    const context = buildContext(entities, diagramType);
    const diff = computeDiffPatch(this.getContext(input.sessionId) ?? null, context);
    this.setContext(input.sessionId, context);

    const contract = applyGuardRails(buildContract(context, diff));
    const systemPrompt = buildSystemPrompt(session.role);

    return { contract, diff, systemPrompt };
  }

  formatDiagram(sessionId: string, format: DiagramFormat): DiagramContract {
    const session = this.getSession(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }
    const previous = this.getContext(sessionId);
    if (!previous) {
      throw new Error('No diagram available for session');
    }
    const context = buildContext(previous.entities, format);
    const diff = computeDiffPatch(previous, context);
    this.setContext(sessionId, context);
    return applyGuardRails(buildContract(context, diff));
  }

  exportSession(sessionId: string): { session: Session; history: MessageInput[]; contract: DiagramContract | null } {
    const session = this.getSession(sessionId);
    if (!session) {
      throw new Error('Session not found');
    }
    const context = this.getContext(sessionId);
    const contract = context
      ? applyGuardRails(buildContract(context, computeDiffPatch(context, context)))
      : null;
    return {
      session,
      history: this.store.getHistory(sessionId),
      contract,
    };
  }
}
