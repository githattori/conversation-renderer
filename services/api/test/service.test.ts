import { describe, expect, it } from 'vitest';
import { DiagramService } from '../src/service.js';

describe('DiagramService', () => {
  it('creates session and generates mermaid DSL with guard rails', () => {
    const service = new DiagramService();
    const session = service.createSession('planner');

    const result = service.processMessage({
      sessionId: session.id,
      content: 'Process order intake -> Validate payment -> Schedule delivery for client alice@example.com',
    });

    expect(result.contract.diagram.format).toBe('mermaid');
    expect(result.contract.diagram.dsl).toMatchSnapshot('mermaid-dsl');
    expect(result.contract.diagram.layout.nodes).toHaveLength(3);
    expect(result.contract.conflicts).toEqual([]);
    expect(JSON.stringify(result.contract.diagram.dsl)).not.toContain('alice@example.com');
    expect(result.diff.addedNodes).toHaveLength(3);
  });

  it('flags isolated nodes and masks PII', () => {
    const service = new DiagramService();
    const session = service.createSession('analyst');

    const result = service.processMessage({
      sessionId: session.id,
      content: 'Escalation contact 555-555-5555',
    });

    expect(result.contract.conflicts.some((conflict) => conflict.includes('isolated'))).toBe(true);
    const layoutLabels = result.contract.diagram.layout.nodes.map((node) => node.label);
    expect(layoutLabels.join(' ')).not.toContain('555');
  });

  it('generates mindmap DSL when brainstorming context is detected', () => {
    const service = new DiagramService();
    const session = service.createSession('facilitator');

    service.processMessage({
      sessionId: session.id,
      content: 'Brainstorm ideas for launch including Marketing Plan, Engineering tasks, Customer Support readiness',
    });

    const exportPayload = service.exportSession(session.id);
    expect(exportPayload.contract?.diagram.format).toBe('mindmap');
    expect(exportPayload.contract?.diagram.dsl).toMatchSnapshot('mindmap-dsl');
  });

  it('allows reformatting a diagram into a different format', () => {
    const service = new DiagramService();
    const session = service.createSession('analyst');

    service.processMessage({
      sessionId: session.id,
      content: 'System components: API Gateway, Auth Service, Data Lake, BI Dashboard',
    });

    const contract = service.formatDiagram(session.id, 'mermaid');
    expect(contract.diagram.format).toBe('mermaid');
    expect(contract.diagram.dsl).toMatchSnapshot('formatted-mermaid-dsl');
  });
});
