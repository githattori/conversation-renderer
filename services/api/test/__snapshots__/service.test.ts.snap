// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DiagramService > allows reformatting a diagram into a different format > formatted-mermaid-dsl 1`] = `
"flowchart LR
node_0[System]
node_1[API Gateway]
node_2[Auth Service]
node_3[Data Lake]
node_4[BI Dashboard]
node_0 --> node_1
node_1 --> node_2
node_2 --> node_3
node_3 --> node_4"
`;

exports[`DiagramService > creates session and generates mermaid DSL with guard rails > mermaid-dsl 1`] = `
"flowchart LR
node_0[Process]
node_1[Validate]
node_2[Schedule]
node_0 --> node_1
node_1 --> node_2"
`;

exports[`DiagramService > generates mindmap DSL when brainstorming context is detected > mindmap-dsl 1`] = `
{
  "edges": [
    {
      "label": null,
      "source": "node_0",
      "target": "node_1",
    },
    {
      "label": null,
      "source": "node_0",
      "target": "node_2",
    },
    {
      "label": null,
      "source": "node_0",
      "target": "node_3",
    },
  ],
  "nodes": [
    {
      "id": "node_0",
      "label": "Brainstorm",
      "level": 0,
    },
    {
      "id": "node_1",
      "label": "Marketing Plan",
      "level": 1,
    },
    {
      "id": "node_2",
      "label": "Engineering",
      "level": 1,
    },
    {
      "id": "node_3",
      "label": "Customer Support",
      "level": 2,
    },
  ],
  "root": "node_0",
}
`;
